### ISSUE TITLE
When I close mintty window, my emacs session lost.

### ISSUE THREAD

**Issue by netjune** [mintty/mintty#1063]
I'd like mintty to send some keys to terminal app, when I close mintty window.  For example, send `C-xC-c` to emacs.
Is it possible?

**Comment by mintty** [comment]
Yes.
Mintty invokes termination by sending a SIGHUP to the child process, then expects the child to terminate, but if it doesn't, nothing happens. So you just need to catch SIGHUP in the child process. My own editor mined uses that to control termination itself.
This is overriden if you hold shift when clicking the exit button, in which case mintty sends SIGKILL.

**Comment by mintty** [comment]
Note this is mentioned in the manual page, section "Closing a session".

**Comment by mintty** [comment]
Actually, my response was about preventing mintty from closing.
If you want it to send characters to the client (and not close), that's also easy to implement.
As this has never been requested before, however, I'd appreciate if you elaborate a bit more on the use case.

**Comment by mintty** [comment]
Please check the uploaded feature, option `ExitCommands` (see manual).

**Comment by netjune** [comment]
> 
> 
> Yes.
> Mintty invokes termination by sending a SIGHUP to the child process, then expects the child to terminate, but if it doesn't, nothing happens. So you just need to catch SIGHUP in the child process. My own editor mined uses that to control termination itself.
> This is overriden if you hold shift when clicking the exit button, in which case mintty sends SIGKILL.

Yes. When I close mintty window, emacs recieves SIGHUP. It just quits. I just want to save all files before quiting. I have not found how to custom this in emacs.



> 
> 
> Please check the uploaded feature, option `ExitCommands` (see manual).

I see it. It is great. Then I can send "^Xs" to emacs to save all modified buffers before quiting.
I can't compile mintty. So I will test it when there is a prebuilt mintty.exe.

Thank you!

**Comment by mintty** [comment]
Compiling mintty is easy, you just need packages gcc-core and make.

**Comment by netjune** [comment]
I have VS2017 installed. Do I need to install MSYS2? Is there a document for building?

**Comment by mintty** [comment]
VS is a Windows thing. Mintty is built in cygwin. I was assuming that you run mintty and emacs from cygwin. If that is not the case, you may have some standalone bundling, so in fact without development environment. Please describe your installation, and I'll upload a binary as an exception.

**Comment by netjune** [comment]
I'm using mintty.exe in git for windows package. I also installed wsltty. Both of them have mintty.exe. I can use either of them.

**Comment by mintty** [comment]
Yes, Git for Windows is a "closed shop" installation with no easy access to further packages.
Note, the attached binary also includes work-in-progress not yet in the repository.
[mintty-1063.zip](https://github.com/mintty/mintty/files/5579464/mintty-1063.zip)

**Comment by netjune** [comment]
> 
> 
> Yes, Git for Windows is a "closed shop" installation with no easy access to further packages.
> Note, the attached binary also includes work-in-progress not yet in the repository.
> [mintty-1063.zip](https://github.com/mintty/mintty/files/5579464/mintty-1063.zip)

I put the the new mintty.exe to wsltty's directory `bin\mintty-d.exe`. It seems that it can't work with the cygwin1.dll in it.

```
      0 [main] ssh (11824) D:\apps\git\usr\bin\ssh.exe: *** fatal error - cygheap base mismatch detected - 0x180347408/0x18034C408.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
```

**Comment by netjune** [comment]
I put the new mintty.exe to wsltty's bin directory(`bin\mintty-d.exe`) and add the following line to `etc/mintty-debug`:
`ExitCommands=bash:^X^S;mined:^[q;emacs:^X^S`

It seems that it doesn't work.

What I do:
1. I run command `bin\mintty-d.exe --WSL= -c "etc\minttyrc-debug"`. 
2. In WSL's bash shell, I run `emacs test.txt`.
3. I modify the buffer content in emacs, and then close mintty window.
4. I open the test.txt again and see that the modification was not saved.

**Comment by mintty** [comment]
This feature _can_ not work in WSL (as mentioned in the manual). It selects an Exit action dependent on the current foreground program. In the case of WSL, that is just the wsl bridge; it is not possible to detect the current foreground program on the WSL side (unless wslbridge2 would provide a gateway mechanism to communicate a respective request and respond to it).

**Comment by netjune** [comment]
> 
> 
> This feature _can_ not work in WSL (as mentioned in the manual). It selects an Exit action dependent on the current foreground program. In the case of WSL, that is just the wsl bridge; it is not possible to detect the current foreground program on the WSL side (unless wslbridge2 would provide a gateway mechanism to communicate a respective request and respond to it).

For this case, can I set a fallback item? For example, `ExitCommands=*:^X^S`
I often open sevral mintty windows, each one of which is dedicated for one emacs only.

**Comment by mintty** [comment]
Released 3.4.4.
